server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1156,57 +1156,62 @@ private class IndexSnapshotsDeletion {
-                                .getNumberOfShards();
-                        } catch (Exception ex) {
-                            logger.warn(
-                                () -> format("[%s] [%s] failed to read metadata for index", indexMetaGeneration, indexId.getName()),
-                                ex
-                            );
-                            // Just invoke the listener without any shard generations to count it down, this index will be cleaned up
-                            // by the stale data cleanup in the end.
-                            // TODO: Getting here means repository corruption. We should find a way of dealing with this instead of just
-                            // ignoring it and letting the cleanup deal with it.
-                            return null;
-                        }
-                    }));
-                }
+            private synchronized void updateShardCount(int newShardCount) {
+                shardCount = Math.max(shardCount, newShardCount);
+            }
 
-                // -------------------------------------------------------------------------------------------------------------------------
-                // Determining the shard count
+            void run(ActionListener<Collection<ShardSnapshotMetaDeleteResult>> shardResultsListener) {
+                determineShardCount(shardResultsListener.delegateFailureAndWrap((l, v) -> processShards(l)));
+            }
 
-                shardCountListener.addListener(deleteIndexMetadataListener.delegateFailureAndWrap((delegate, counts) -> {
-                    final int shardCount = counts.stream().mapToInt(i -> i).max().orElse(0);
-                    if (shardCount == 0) {
-                        delegate.onResponse(null);
-                        return;
+            // -----------------------------------------------------------------------------------------------------------------------------
+            // Determining the shard count by reading all the relevant index metadata
+
+            private void determineShardCount(ActionListener<Void> listener) {
+                try (var listeners = new RefCountingListener(listener)) {
+                    for (final var indexMetaGeneration : snapshotIds.stream()
+                        .filter(snapshotsWithIndex::contains)
+                        .map(id -> originalRepositoryData.indexMetaDataGenerations().indexMetaBlobId(id, indexId))
+                        .collect(Collectors.toSet())) {
+                        // NB since 7.9.0 we deduplicate index metadata blobs, and one of the components of the deduplication key is the
+                        // index UUID; the shard count is going to be the same for all metadata with the same index UUID, so it is
+                        // unnecessary to read multiple metadata blobs corresponding to the same index UUID.
```

Comment:
Not really relevant to reviewing this refactoring. But I need get better understanding on this comment here.

Reply:
See e.g. `org.elasticsearch.repositories.IndexMetaDataGenerations#buildUniqueIdentifier`

Reply:
Just took a closer look into this and I think this `Math.max` operation is still necessary when the index has the same name but different number of shards. For example, an index `songs` goes through "creation with 1 shard -> deletion -> recreation with 3 shards". It gets snapshotted in both the initial creation and the later recreation. 

If we delete both snapshots in a single call, the `indexMetaGenerations` [computed here](https://github.com/elastic/elasticsearch/blob/de09130042f6a7e7f4296a70874c6042865ed471/server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java#L1160-L1163) will contain two generations and subsequently lead to shard counts to be `[1, 3]` and we need max between them to determine the final outcome of `3`.

Reply:
Yes that's right, these things are combined by index _name_ so it is indeed possible to see varying shard counts, and we need to take the max to account for that. The point is that there may be several metadata blobs corresponding to the same index UUID (e.g. reflecting variations in mappings or settings between different snapshots) but there's no need to read more than one if we only care about the shard count.

Reply:
Ok right. Yes for the same index UUID (the one from the cluster), the shard count must be the same so that we can indeed save some work here. 👍